From: Christoph Reiter <reiter.christoph@gmail.com>
Subject: [PATCH] win32: Work around a SpiderMonkey shutdown deadlock

JS_ShutDown() in DLL_THREAD_DETACH deadlocks under the Windows loader lock.
Move it to CLI cleanup as a workaround; I do not know the correct fix.

diff --git a/gjs/engine.cpp b/gjs/engine.cpp
--- a/gjs/engine.cpp
+++ b/gjs/engine.cpp
@@ -113,10 +113,6 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved) {
             gjs_is_inited = true;
         } break;
 
-        case DLL_THREAD_DETACH:
-            JS_ShutDown();
-            break;
-
         default: {
         }
     }
diff --git a/gjs/console.cpp b/gjs/console.cpp
--- a/gjs/console.cpp
+++ b/gjs/console.cpp
@@ -19,6 +19,10 @@
 #include <glib-object.h>
 #include <glib.h>
 
+#ifdef G_OS_WIN32
+#    include <js/Initialization.h>
+#endif
+
 #include "gjs/auto.h"
 #include "gjs/gerror-result.h"
 #include "gjs/gjs.h"
@@ -230,6 +234,12 @@ class TraceFD {
 };
 
 int main(int argc, char** argv) {
+#ifdef G_OS_WIN32
+    struct EngineShutdown {
+        ~EngineShutdown() { JS_ShutDown(); }
+    } engine_shutdown;
+#endif
+
     Gjs::AutoError error;
     int gjs_argc = argc;
     bool interactive_mode = false;
